pkg/compactor/visit_marker.go

```diff
@@ -0,0 +1,194 @@
+package compactor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/thanos/pkg/compact"
+
+	"github.com/cortexproject/cortex/pkg/util/runutil"
+)
+
+var (
+	errorVisitMarkerNotFound  = errors.New("visit marker not found")
+	errorUnmarshalVisitMarker = errors.New("unmarshal visit marker JSON")
+)
+
+type VisitStatus string
+
+const (
+	Pending    VisitStatus = "pending"
+	InProgress VisitStatus = "inProgress"
+	Completed  VisitStatus = "completed"
+	Failed     VisitStatus = "failed"
+)
+
+type VisitMarker interface {
+	GetVisitMarkerFilePath() string
+	UpdateStatus(ownerIdentifier string, status VisitStatus)
+	GetStatus() VisitStatus
+	IsExpired(visitMarkerTimeout time.Duration) bool
+	String() string
+}
+
+type VisitMarkerManager struct {
+	bkt                    objstore.InstrumentedBucket
+	logger                 log.Logger
+	ownerIdentifier        string
+	visitMarker            VisitMarker
+	visitMarkerReadFailed  prometheus.Counter
+	visitMarkerWriteFailed prometheus.Counter
+}
+
+func NewVisitMarkerManager(
+	bkt objstore.InstrumentedBucket,
+	logger log.Logger,
+	ownerIdentifier string,
+	visitMarker VisitMarker,
+	visitMarkerReadFailed prometheus.Counter,
+	visitMarkerWriteFailed prometheus.Counter,
+) *VisitMarkerManager {
+	return &VisitMarkerManager{
+		bkt:                    bkt,
+		logger:                 log.With(logger, "type", fmt.Sprintf("%T", visitMarker)),
+		ownerIdentifier:        ownerIdentifier,
+		visitMarker:            visitMarker,
+		visitMarkerReadFailed:  visitMarkerReadFailed,
+		visitMarkerWriteFailed: visitMarkerWriteFailed,
+	}
+}
+
+func (v *VisitMarkerManager) HeartBeat(ctx context.Context, errChan <-chan error, visitMarkerFileUpdateInterval time.Duration, deleteOnExit bool) {
```

Comment:
I am a little bit worried about the logs in the heartbeat logic. Are they worth it as `Info` level or fine to be `Debug`?

Reply:
Do we need to expose `deleteOnExit`? I see we set it to true for all 2 usage of this function

Reply:
Those logs in heartbeat only logged at start and end of the heartbeat. Would this still be concern?

For `deleteOnExit`, this adds flexibility for other type of visit marker to retain the file once on exit. For example, we can rewrite block visit marker to use this base class. Also, partition visit marker in next couple PRs could utilize it.

Reply:
I am good thanks Alex